### PR TITLE
Update(orderBy): Guaranteed stable sort

### PR DESCRIPTION
**Pull request by BobChao87**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

Feature

**What is the current behavior? (You can also link to an open issue here)**

#14881 

**What is the new behavior (if this is a feature change)?**

Falls back on built-in comparison function if user-provided comparison function fails to break a tie between elements being ordered.

**Does this PR introduce a breaking change?**

Yes

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

BREAKING CHANGE: Using `orderBy` with a custom `comparator` function while
simultaneously setting `reverse` to `true` results in a potentially different
list order than previously.

Specifically, when the custom `comparator` fails to sort two objects (compares
them as being "equal"), the objects would previously have remained in their
original order in the array relative to each other. They will now switch places
as the `reverse` will be more accurately respected.

To migrate, make sure that your custom `comparator` properly differentiates
between all cases that your code base cares about and that "equal" cases are
not order dependent. We expect that due to the nature of this change, since the
entries are considered "equal", that very few projects will be affected by this
change in all but a minor visual manor that still respects the expected
operation being performed.

### CODE REVIEW

File: test/ng/filter/orderBySpec.js [hunk 1]
```diff
@@ -457,6 +457,21 @@ describe('Filter: orderBy', function() {
...
+        // Some list that won't be sorted "naturally", i.e. should sort to ['a', 'B', 'c']
+        var items = ['c', 'a', 'B'];
+        var expr = null;
+        function comparator() {
+          return 0;
+        }
+        var reversed = ['B', 'a', 'c'];
+        var sorted = ['a', 'B', 'c'];
+
+        expect(orderBy(items)).toEqual(sorted);
```
Comment (gkalpak) [inline test/ng/filter/orderBySpec.js]: This doesn't seem related.

File: src/ng/filter/orderBy.js [hunk 2]
```diff
@@ -115,7 +115,7 @@
  *
  * @param {boolean=} reverse - If `true`, reverse the sorting order.
  * @param {(Function)=} comparator - The comparator function used to determine the relative order of
- *    value pairs. If omitted, the built-in comparator will be used.
+ *    value pairs. If omitted or finds two entries to be equal, the built-in comparator will be used.
```
Comment (gkalpak) [inline src/ng/filter/orderBy.js]: I think thisis more confusing than helpful. If they are interested thay read the details below.

**Review by gkalpak — CHANGES_REQUESTED** [review]
A couple of minor comments only.
I am not sure about the BC tbh. I don't really consider it to be one. It would only break usecases where all of the following conditions are true:

- The app is using a custom comparator.
- The custom comparator fails to differentiate between two values.
- The app expects both ascending and descending sorting to yield the same results (which depend on the browser).

**Comment by BobChao87** [comment]
Fair enough about the breaking change. I wasn't sure either, but figured it would be better to write it up than gloss over it. 

I'll get right to work on those changes.

File: src/ng/filter/orderBy.js [hunk 3]
```diff
@@ -599,7 +602,7 @@ function orderByFilter($parse) {
         }
       }
 
-      return compare(v1.tieBreaker, v2.tieBreaker) * descending;
+      return (compare(v1.tieBreaker, v2.tieBreaker) || defaultCompare(v1.tieBreaker, v2.tieBreaker)) * descending;
```
Comment (lgalfaso) [inline src/ng/filter/orderBy.js]: I do not fully understand this change, this should be just
`return defaultCompare(v1.tieBreaker, v2.tieBreaker) * descending;`

There is no reason for an external function to understand the internal `tieBreaker` structure.

Comment (BobChao87) [inline src/ng/filter/orderBy.js]: This is not intended to defer to the `defaultCompare` function over the user-provided function, but rather to provide a reasonable fallback should a user-provided function fail to the break the tie. See [#14481 (comment)](https://github.com/angular/angular.js/issues/14881#issuecomment-231164555)

Comment (gkalpak) [inline src/ng/filter/orderBy.js]: @lgalfaso, for better or worse, this tiebreaker structure is documented and has been used with custom comparator 😕 Do you think we should change that?
I thought it is useful to give the custom comparator a chance to break the tie in its own way (in the rare case where this will be needed).

**Comment by lgalfaso** [comment]
Was not aware that this was added as part of the public API. Once this was
added, then I think it is best to keep it.

**Comment by BobChao87** [comment]
Was there anything else pending or can we go ahead and mark this as ready to be merged? I think we've covered all the concerns that have been raised.

**Review by gkalpak — APPROVED** [review]

File: src/ng/filter/orderBy.js [hunk 4]
```diff
@@ -460,6 +460,9 @@
  * way. (When specifying a custom comparator, you also need to pass a value for the `reverse`
  * argument - passing `false` retains the default sorting order, i.e. ascending.)
  *
+ * If your comparator fails to differentiate between two items, they will fall back to the built-in
+ * comparator function, this produces a guarantee of stable sorting.
+ *
```
Comment (gkalpak) [inline src/ng/filter/orderBy.js]: Actually, this is the wrong place for this comment. It should be moved to [L55][1] and be changed to something like:

> If a custom comparator cannot still can't distinguish between two items, then they will be sorted based on their index using the built-in comparator.

[1]: https://github.com/BobChao87/angular.js/blob/8f236af514d749f194c0a1ef8fa7f692386997eb/src/ng/filter/orderBy.js#L55

**Review by gkalpak — CHANGES_REQUESTED** [review]
Can you also change the commit message to adhere to our guidelines. E.g.:

> fix(orderByFilter): guarantee stable sort
>
> If the user-provided comparator fails to break a tie, fall back to the built-in comparator.
>
> Fixes #14881

**Comment by BobChao87** [comment]
Apologies about the commit message, apparently it's been a while since I read the contributing guidelines and I use a similar scheme on another project.

File: src/ng/filter/orderBy.js [hunk 5]
```diff
@@ -49,9 +49,8 @@
  * second, or `1` otherwise.
  *
  * In order to ensure that the sorting will be deterministic across platforms, if none of the
- * specified predicates can distinguish between two items, `orderBy` will automatically introduce a
- * dummy predicate that returns the item's index as `value`.
- * (If you are using a custom comparator, make sure it can handle this predicate as well.)
```
Comment (gkalpak) [inline src/ng/filter/orderBy.js]: This line should not be removed. It is still necessary for the comparator to handle the tie-break value. E.g. if it has an array of strings, it should not expect that all values will be strings, because there is the possibility of getting passed a number (the index).

File: src/ng/filter/orderBy.js [hunk 6]
```diff
@@ -49,9 +49,8 @@
...
+ * specified predicates nor the custom comparator can distinguish between two items, `orderBy` will
```
Comment (gkalpak) [inline src/ng/filter/orderBy.js]: This change is confusing imo. The custom comparator is not something additional to the predicates. They are used together (the predicates provide the values and the comparator compares them).

**Review by gkalpak — APPROVED** [review]
A couple of nit. I will fix up while merging.
